### PR TITLE
Allow fake players to be given stages. 

### PR DESCRIPTION
**Pull request by cpw**
This should mean that gamestages are much more compatible with automated
miners and such.

You could probably provide a config to whitelist specific fakeplayer ids
rather than all. The only call that won't work is "getAllStages" since
that's not a globally stored list.

This would likely address issues https://github.com/Darkosto/SevTech-Ages/issues/1733 and https://github.com/Darkosto/SevTech-Ages/issues/1803

### CODE REVIEW

**Comment by cpw** [comment]
My opinion is that gamestages should not apply to fake players. They're not really exposed through the usual mechanisms, you can't interact with them with commands for example.

You could perhaps have two policies "all" and "none" - and fakeplayers could be swapped from one policy to another if you wanted. Trying to restrict fakeplayers to the same system players have would require a significant rewrite of how fakeplayers themselves work, and it would result in a lot of weirdness and confusion.

**Comment by AlexIIL** [comment]
Sorry, @justinrusso asked for my recommendation on this and I never actually replied to https://github.com/Darkosto/SevTech-Ages/issues/1733 so this seems like a good place.

I'd recommend separating player game data from the EntityPlayer itself, and use a separate system for saving and loading. Obviously it's quite a large change to move save data from player entity to somewhere else, but it should be possible. This would be useful for the case where the machine stores the `UUID`(or `GameProfile`) of the player who placed the block, like with [build](https://github.com/BuildCraft/BuildCraft/blob/8.0.x/common/buildcraft/lib/tile/TileBC_Neptune.java#L114)[craft](https://github.com/BuildCraft/BuildCraft/blob/8.0.x/common/buildcraft/lib/tile/TileBC_Neptune.java#L248).

This doesn't work for mods that don't know the player who placed them (or which don't store it yet), however I think that it would still be a useful compromise. It's not a lot of data to just store the UUID of the placing player (and I think this is recommended for forge's permission system?). You can always fall-back on accepting or always denying actions, but it might break progression a bit if one automated miner can mine *everything* and another one requires the player to go through the stages.

Basically, I'd only recommend merging this if moving save files over (or looking up the real, currently loaded, `EntityPlayer` instance and inspecting that) isn't a valid alternative.

**Comment by cpw** [comment]
So, @AlexIIL the problem is that player context is generally not available to automated machines, and few actually support storing and propagating it. The primary purpose of gamestages is to give a progression to _real_ players, not buildcraft quarries. The progression element should be access to the quarry machine, not what the quarry can do, in my opinion.

**Comment by AlexIIL** [comment]
> few actually support storing and propagating it.

This is a very small amount of data to store, and adding support for it really shouldn't be tricky.

>  The progression element should be access to the quarry machine, not what the quarry can do, in my opinion.

**Comment by justinrusso** [comment]
> It's not a lot of data to just store the UUID of the placing player (and I think this is recommended for forge's permission system?)

I am all for getting mods to follow standards. If this is true and is a recommended thing mods should be doing, I am not against complaining to mod devs to add that support of they are not.

In RecipeStages we currently use a `setContainerStage` method that defines set stages that container will support. From there, when the player attempts to craft something, it will check if the player has that stage and only show an output if true. I feel like comparing between the two, @AlexIIL 's method to allow the miner to adapt would be more ideal than just a bypass switch as this would allow players to obtain blocks they shouldn't because of that bypass.

**Comment by AlexIIL** [comment]
> If this is true and is a recommended thing mods should be doing, I am not against complaining to mod devs to add that support of they are not.

I'm originally referring to *something* that I *think* Lex mentioned in the various permission API PR's and issues. However until I can find the source (and I haven't yet) I don't think it can be considered a standard or official recommendation. Sorry. However it still provides one way for mods to work with this, so it's not a complete waste.

**Comment by cpw** [comment]
So I updated my commit after talking to darkosto about how he thinks this should work.

It is basically a fixed policy for each fakeplayer - they get a list of gamestages permanently.

**Comment by AlexIIL** [comment]
That's a good middle ground, but it probably won't handle Buildcraft type of game profile saving very well - as for us, the player name is used rather than sometime like "[buildcraft]". (Which will just fallback to the default).

That being said I probably won't be implementing my own suggestion as a PR.

**Comment by cpw** [comment]
The fakeplayer name is just a string - and it's the same field as the player's name. the rftools builder for example has the name "rftools_builder".

so you'd have an entry in the config file with that name, and gamestages to apply to the fakeplayer.

If you don't want commands to interact with the fakeplayers (they're not really visible to the commands at present, because fakeplayers are never actually *on* the server's playerlist) just leave it as is.

**Comment by darkhax** [comment]
Thanks for making this PR and starting this discussion. I have been planning to do a rewrite to improve the internals for a while now. I opened up [this](https://github.com/Darkhax-Minecraft/Game-Stages/issues/15) issue so people can discuss the rewrite in one place. 

As for this PR, I think this is a good solution without breaking compatibility although I am curious to @AlexIIL 's response. There should be some way to identify a Buildcraft FakePlayer from a normal one, perhaps by class name?

**Comment by AlexIIL** [comment]
@darkhax this PR perfectly addresses the case that cpw is talking about (as mods don't give enough information to work with), and #15 should address Buildcraft's case properly. So I don't recommend trying to identify the Buildcraft fake player (or any others that same the full player name) alongside this.

(Although yes, currently you can identify the Buildcraft fake player by [class name](https://github.com/BuildCraft/BuildCraft/blob/8.0.x/common/buildcraft/lib/fake/FakePlayerBC.java) - however this was only done to override some entity player breaking methods that should be in forge, so I doubt we will keep if forever).
